packages/pigeon/lib/pigeon_lib.dart

```diff
@@ -625,37 +640,57 @@ class _RootBuilder extends dart_ast_visitor.RecursiveAstVisitor<Object?> {
         );
       }
     } else {
-      _currentClass = Class(name: node.name.name, fields: <Field>[]);
+      _currentClass = Class(name: node.name.name, fields: <NamedType>[]);
     }
 
     node.visitChildren(this);
     return null;
   }
 
+  NamedType formalParameterToField(dart_ast.FormalParameter parameter) {
+    final dart_ast.TypeName typeName = parameter.childEntities.firstWhere(
+        (dart_ast_syntactic_entity.SyntacticEntity e) =>
+            e is dart_ast.TypeName) as dart_ast.TypeName;
+    final String argType = typeName.name.name;
+    final bool isNullable = typeName.question != null;
+    final List<TypeDeclaration>? argTypeArguments =
+        typeAnnotationsToTypeArguments(typeName.typeArguments);
+    return NamedType(
+        type: TypeDeclaration(
+            dataType: argType,
+            isNullable: isNullable,
+            typeArguments: argTypeArguments),
+        name: parameter.identifier?.name ?? '',
+        offset: null);
+  }
+
+  static T? getFirstChildOfType<T>(dart_ast.AstNode entity) {
+    for (final dart_ast_syntactic_entity.SyntacticEntity child
+        in entity.childEntities) {
+      if (child is T) {
+        return child as T;
+      }
+    }
+    return null;
+  }
+
   @override
   Object? visitMethodDeclaration(dart_ast.MethodDeclaration node) {
     final dart_ast.FormalParameterList parameters = node.parameters!;
-    late String argType;
-    bool isNullable = false;
-    if (parameters.parameters.isEmpty) {
-      argType = 'void';
-    } else {
-      final dart_ast.FormalParameter firstParameter =
-          parameters.parameters.first;
-      final dart_ast.TypeName typeName = firstParameter.childEntities
-          // ignore: always_specify_types
-          .firstWhere((e) => e is dart_ast.TypeName) as dart_ast.TypeName;
-      argType = typeName.name.name;
-      isNullable = typeName.question != null;
-    }
+    final List<NamedType> arguments =
+        parameters.parameters.map(formalParameterToField).toList();
     final bool isAsynchronous = _hasMetadata(node.metadata, 'async');
     if (_currentApi != null) {
       _currentApi!.methods.add(Method(
           name: node.name.name,
-          returnType: node.returnType.toString(),
-          argType: argType,
-          isReturnNullable: node.returnType!.question != null,
-          isArgNullable: isNullable,
+          returnType: TypeDeclaration(
+              dataType: getFirstChildOfType<dart_ast.SimpleIdentifier>(
+                      node.returnType!)!
```

Comment:
Why are these `!`s (and those just belowe) safe? Needs a comment.

Reply:
Done, I didn't even know you could create methods without named return types.  I guess you can:
```dart
printInt(int x) {
  print(x);
}
```

That's kind of goofy in a 1980s C kind of way goofy.